openwisp_notifications/base/models.py

```diff
@@ -277,14 +280,45 @@ class Meta:
         ]
 
     def __str__(self):
-        return '{type} - {organization}'.format(
-            type=self.type_config['verbose_name'],
-            organization=self.organization,
-        )
+        type_name = self.type_config.get('verbose_name', 'Global Setting')
+        if self.organization:
+            return '{type} - {organization}'.format(
+                type=type_name,
+                organization=self.organization,
+            )
+        else:
+            return type_name
+
+    def validate_global_setting(self):
+        if self.organization is None and self.type is None:
+            if (
+                self.__class__.objects.filter(
+                    user=self.user,
+                    organization=None,
+                    type=None,
+                )
+                .exclude(pk=self.pk)
+                .exists()
+            ):
+                raise ValidationError("There can only be one global setting per user.")
 
     def save(self, *args, **kwargs):
+        self.validate_global_setting()
         if not self.web_notification:
             self.email = self.web_notification
+        if not self.organization and not self.type:
+            try:
+                original = self.__class__.objects.get(pk=self.pk)
+                if self.web and (self.email == original.email):
+                    self.user.notificationsetting_set.exclude(pk=self.pk).update(
+                        web=self.web
+                    )
+                else:
+                    self.user.notificationsetting_set.exclude(pk=self.pk).update(
+                        web=self.web, email=self.email
+                    )
+            except self.__class__.DoesNotExist:
+                pass
         return super().save(*args, **kwargs)
```

Comment:
What are the drawbacks for upgrading the NotificationSetting using `update()` 

```python
self.user.notificationsetting_set(web=self.web, email=self.email)
```

What case would be left uncovered? Can you demonstrate that with a failing test case?

Reply:
I have added a failing testcase for this 

https://github.com/openwisp/openwisp-notifications/blob/a796372fac040afdda7b314f4d491f6eb6167e71/openwisp_notifications/tests/test_api.py#L850-L870

Also did add for the global setting here

https://github.com/openwisp/openwisp-notifications/blob/a796372fac040afdda7b314f4d491f6eb6167e71/openwisp_notifications/tests/test_notification_setting.py#L292-L307

Let me know what do you think about this.

Reply:
I commented out the code in question, but the added test didn't fail. :(

I was expecting the test to fail without the added logic.